### PR TITLE
[SPARK-12252][SPARK-12131][SQL] refactor MapObjects to make it less hacky

### PR DESCRIPTION
**Pull request by cloud-fan**
in https://github.com/apache/spark/pull/10133 we found that, we shoud ensure the children of `TreeNode` are all accessible in the `productIterator`, or the behavior will be very confusing.

In this PR, I try to fix this problem by expsing the `loopVar`.

This also fixes SPARK-12131 which is caused by the hacky `MapObjects`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala [hunk 1]
```diff
@@ -323,22 +323,26 @@ case class WrapOption(child: Expression)
...
-  override def nullable: Boolean = false
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported.")
+object LoopVar {
+  private val curId = new java.util.concurrent.atomic.AtomicInteger()
 
+  def apply(dataType: DataType): LoopVar = {
+    val loopValue = "MapObjects_loopValue" + curId.getAndIncrement()
+    val loopIsNull = "MapObjects_loopIsNull" + curId.getAndIncrement()
+    LoopVar(loopValue, loopIsNull, dataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala]: Here is a little hacky, but I wanna argue that it's way less hacky than the previous implementation and I make the variable name verbose to make it less possible to get conflicted.

**Comment by cloud-fan** [comment]
cc @liancheng @marmbrus

**Comment by liancheng** [comment]
This does look much better. Can we add a `MapObjects.apply` to hide `LoopVar` so that users don't need to create them manually?

**Comment by marmbrus** [comment]
This seems fine, but I agree we should hide the construction of the loop variable.  There is nothing wrong with the apply taking a `Expression => Expression`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala [hunk 2]
```diff
@@ -326,19 +326,28 @@ case class WrapOption(child: Expression)
...
+object MapObjects {
+  private val curId = new java.util.concurrent.atomic.AtomicInteger()
 
+  def apply(
+      function: Expression => Expression,
+      inputData: Expression,
+      elementType: DataType): MapObjects = {
+    val loopValue = "MapObjects_loopValue" + curId.getAndIncrement()
+    val loopIsNull = "MapObjects_loopIsNull" + curId.getAndIncrement()
+    val loopVar = LambdaVariable(loopValue, loopIsNull, elementType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala]: Here is a little hacky, but I wanna argue that it's way less hacky than the previous implementation and I make the variable name verbose to make it less possible to get conflicted.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala [hunk 3]
```diff
@@ -326,19 +326,28 @@ case class WrapOption(child: Expression)
  * A place holder for the loop variable used in [[MapObjects]].  This should never be constructed
  * manually, but will instead be passed into the provided lambda function.
  */
-case class LambdaVariable(value: String, isNull: String, dataType: DataType) extends Expression {
+case class LambdaVariable(value: String, isNull: String, dataType: DataType) extends LeafExpression
+  with Unevaluable {
 
-  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String =
-    throw new UnsupportedOperationException("Only calling gen() is supported.")
+  override def nullable: Boolean = true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala]: I think the `loopVar` is nullable?

**Comment by liancheng** [comment]
Great, so the inner class issue is also resolved. I'm closing #10133 then.

**Comment by gatorsmile** [comment]
Regarding the JIRA SPARK-12131, how about the following statement? 
```scala
implicit val kryoEncoder = Encoders.kryo[KryoClassData]
val ds = Seq((KryoClassData("a", 1), KryoClassData("b", 2)).toDS()
```
This does not work. @liancheng just closed this PR, which sounds a similar issue for tuple, right? Will this PR resolves this issue? @cloud-fan

**Comment by cloud-fan** [comment]
@gatorsmile , the problem is: currently kryo/javaSerialization encoder is not composable, so it's another story, feel free to open a JIRA for it.

**Comment by gatorsmile** [comment]
@cloud-fan Sure, let me open it now. Thank you!

**Comment by liancheng** [comment]
@cloud-fan Could you please mention SPARK-12131 in the PR title and description since it's also fixed in this PR? Then I can merge it to master.

**Comment by liancheng** [comment]
Thanks, merging to master.

**Pull request closed without merging**